### ISSUE TITLE
Implement Omemo / Axolotl

### ISSUE THREAD

**Issue by mrbiber** [ChatSecure/ChatSecure-iOS#376]
It would be so great if ChatSecure supported Conversation's [Omemo](http://conversations.im/omemo/) protocol for multi-party, multi-device OTR.

**Comment by chrisballinger** [comment]
It's on the roadmap! 🎉

**Comment by tristan-k** [comment]
> Soon we will be implementing OMEMO Encryption in ChatSecure iOS, and immediately contributing our OMEMO XEP code upstream to XMPPFramework so that other apps can benefit. We plan to utilize the pre-existing Objective-C library AxolotlKit, written by Frederic Jacobs, that has been used in production since the release of Open Whisper System’s Signal v2.0 for iOS. Unfortunately AxolotlKit is still currently GPL (and therefore not redistributable to the App Store) so this work is on hold until we can negotiate a change to an App Store-comptible copyleft license like LGPLv2 or MPL 2.0 from Fred and Moxie.

> Written by [Chris Ballinger](https://chatsecure.org/blog/chatsecure-conversations-zom/) — OCTOBER 02, 2015

Any update on this matter? Even though AxolotlKit is not redistributable in the App Store, why not make it possible to [sideload](http://osxdaily.com/2016/01/12/howto-sideload-apps-iphone-ipad-xcode/) the app for the more tech-savvy people?

**Comment by therob84** [comment]
I am also waiting very eagerly that "Conversations" will get another app which supports OMEMO/Axolotl over XMPP.

Any news about this for chatsecureZOM (esp. for iOS) as the last official announcement is from [october 2015](https://chatsecure.org/blog/chatsecure-conversations-zom/)

**Comment by chrisballinger** [comment]
This work is semi-permanently on hold because of the license conflict. Moxie said that the public specification for Axolotl is incomplete, so it will be impossible for us to produce an alternative implementation that isn't a derivative work of one of the GPL libraries.

**Comment by amenk** [comment]
Can  you explain? Isn't this thing here GPL? And why can't you use their axolotl implementation?

**Comment by amenk** [comment]
Okay. I read above that AppStore does not allow GPL.. Sorry for the noise

**Comment by therob84** [comment]
Thanks for clarification, Chris!
Although I like the idea that AppleIstore is/do not allowed to use GPL software, in this special case it is a pain, that the great OMEMO (Axolotl) protocol cannot be ported to any of the iOS Apps.....and thus is been hindered to become a more widespread alternative for XMPP...at least for iPhone users...which I unfortunately have in my circle of friends...

@chrisballinger Do you think there will be a breakthrough in porting OMEMO to any iOS App in the next....months? Or should we lay it to the graves?

btw: What I don't understand: e.g. Whattsapp also uses Axolotl encryption of TextSecure/Signal/OpenWhisperSystem, which also should be licensed under GPL, but is allowed in the AppleStore? 
Where is the difference? [Here ](https://chatsecure.org/blog/chatsecure-core/)are some more infos about, but not that I do understand it in detail :-/

**Comment by amenk** [comment]
@therob84 It thinks WhatsApp does not encrypt end-to-end when chatting with iOS users
https://www.quora.com/Is-WhatsApp-encrypted-for-iPhone

**Comment by therob84** [comment]
OK, for Whattsapp this could be true.

But what's about Signal/OpenWhisperSystem in general, which I think uses
E2E encryption at iOS?
https://whispersystems.org/blog/the-new-signal/
It is stated for iOS that also group messages are E2Eenc, which means
they have to use Axolotl, right? (Is here a difference in XMPP and
non-XMPP in this legal-stuff?)

**Comment by chrisballinger** [comment]
Open Whisper Systems owns the full copyright on AxolotlKit so they can relicense it for distribution on the App Store for their own apps. They are currently licensing libaxolotl-java to WhatsApp for the Android version, but for whatever reason haven't yet done the same for WhatsApp iOS and AxolotlKit. I've been told there are no near-term plans to license AxolotlKit to other apps.

However, there may be a light at the end of the tunnel: https://github.com/SilentCircle/libsalamander

It appears that the Silent Circle team has implemented their own version of Axolotl using only the public specification and (presumably) avoided any reverse engineering of the GPL code. It is licensed Apache 2.0 so it could be used without issue on the App Store. I'm not sure if the key exchange is compatible with libaxolotl-java, among other things, so there is a chance it may not be compatible with Conversations current implementation of OMEMO.

**Comment by therob84** [comment]
Thanks Chris for your kind and detailed reply.
So for non-programming users do you see anything which could be done to accelerate this process?

**Comment by therob84** [comment]
@chrisballinger: Can you provide us with any progress or promising news as an easter surprise about the OMEMO-topic in ChatSecure for iOS or about libsalamander?

As I can't convice quite some people to use ChatSecure without supporting OMEMO under iOS, I find it a more and more urgent issue...Would be glad to read arguments for holding up hope....
Have nice eastern, cheers!

**Comment by dxerw** [comment]
Damn the licensing..

**Comment by dxerw** [comment]
https://github.com/SilentCircle/libzina

**Comment by therob84** [comment]
@dxerw .... can you comment on this in any way? Is it connnected with libsalamander?
Does it bring us nearer to OMEMO@iOS while beeing compatibel witht OMEMO@Conversations@Android?

Would be great, but I have not enough information on this to be near-term-optimist.......

**Comment by chrisballinger** [comment]
@therob84 It is literally impossible to make something compatible with the current OMEMO spec due to Open Whisper System's decision about licensing SignalProtocolKit. Even if we use another library implementing the Axolotl ratchet, the details of each implementation's protocol and handshake are different and incompatible.

**Comment by chrisballinger** [comment]
I just had an absolutely crazy idea that could potentially get around this issue. Legally it will probably not hold up in court, so this is more of a thought experiment.

The main issue with distributing other people's GPL code on the App Store is GPL section 6 which says "You may not impose any further restrictions on the recipients' exercise of the rights granted herein". The App Store imposes further restrictions regardless of a developer wants them in their release.

The only way to distribute this code without violating the GPL via Apple's restrictions would be to download an external JavaScript AxolotlV3 library on first launch and execute it via JavaScriptCore. The source code distribution would be a free download from an external server containing no restrictions on use. Regardless how the "linking" step is interpreted when running GPL code in a JavaScript interpreter that bridges to Obj-C, it will be the end user doing the linking after the original App Store binary distribution has occurred. The resulting combination cannot be redistributed, but that will fall on the end user and not the app distributor.

This would be similar to how proprietary programs can be used with GPL plugins as long as they are distributed separately and linked by the end user.

**Comment by therob84** [comment]
thanks @chrisballinger for your statement. 

I still hope the best and very welcome the (just) started inter-app-discussion about OMEMO (https://github.com/anurodhp/Monal/issues/9) ... triggering long enaugh from all sides finally lead to the long needed (public visibly) teamwork between you all, 
@chrisballinger ([ChatSecure](https://github.com/ChatSecure/ChatSecure-iOS/)), 
@iNPUTmice ([Conversations](https://github.com/siacs/Conversations/)), 
@anurodhp ([Monal](https://github.com/anurodhp/Monal/)) .....
Keep on this track!

**Comment by the-solipsist** [comment]
@chrisballinger I don't see why your "crazy idea" is all that crazy. I think it circumvents the legal incompatibility rather cleverly.

**Comment by chrisballinger** [comment]
It seems like our funder will not allow us to implement any of these.

**Comment by therob84** [comment]
@chrisballinger: what? I can't beleive that your funder want to miss the (in my eyes) most promising development of the last time in pushing XMPP forward for wide-spread using in terms of security...? 
You are just joking, right?!

So there is no light at the end of the tunnel to give me a chance to use XMPP with my iOS friends? 
...Very sad....
But thanks for beeing clear about that...

**Comment by chrisballinger** [comment]
Moxie very recently told me that he doesn't care if we distribute on the
App Store as long as we otherwise comply with the GPL, but I need something
more concrete for the funder before we can move forward.

**Comment by masvil** [comment]
@chrisballinger please make the funder change his/her mind!

**Comment by austin987** [comment]
I'm an android user (from #740), and would greatly like to see progress on this. Especially given that Android does not suffer from the GPL incompatibility that IOS does, I find it very sad that IOS blocks Android from a more secure and functional chat system.

**Comment by amenk** [comment]
> Moxie very recently told me that he doesn't care if we distribute on the
> App Store as long as we otherwise comply with the GPL, but I need something
> more concrete for the funder before we can move forward.

So Moxie has to relicense the code, right? Did anybody formally approach him?

**Comment by herbsmn** [comment]
@amenk things are looking good! read this thread. https://twitter.com/ChatSecure/status/735930676961148928

**Comment by chrisballinger** [comment]
Although it sounded promising at first, there is ~0% chance I can get the legal stuff sorted out w/ Moxie. We plan to move forward to Matrix's Olm and support them for their audit.

**Comment by herbsmn** [comment]
Sorry to hear that @chrisballinger. Is the reasoning outlined somewhere? It did sound promising.

**Comment by herbsmn** [comment]
![serveimage](https://cloud.githubusercontent.com/assets/8562898/15942703/089dbb2c-2e75-11e6-8559-8e4891699db0.jpeg)
Let's do this!

**Comment by herbsmn** [comment]
Hey @fowlslegs I remember you mentioning that you worked close to OWS. It might be too late, but moxie sounded down with doing a GPL with "additional permissions" that would allow some of their Signal code onto the iOS app store, but now isn't responding to inquiries about proceeding forward. https://twitter.com/ChatSecure/status/735930676961148928 Any chance you can chat with the OWS peeps about it?

**Comment by chrisballinger** [comment]
At this point I don't think it's worth the headache and legal expense (on both sides) to figure this out with OWS because they aren't really interested. It's a lot of work on their part for no gain, and could potentially threaten their business model if the wording isn't correct. I'd rather move forward with Olm once it's audited.

**Comment by fowlslegs** [comment]
So, first I don't think I could sway Moxie and agree that

> I don't think it's worth the headache and legal expense (on both sides) to figure this out with OWS because they aren't really interested

Second, why in particular Olm when there seem to be perhaps more well-known/audited (at least informally) alternative implementations of the double ratchet written in actually safe languages (i.e., not C++, a horrible offender)? Most of these implementations seem to be GPL as well, but it's probable the authors would be more friendly with a GPL modification to allow for App Store compatibility than Moxie has been.

Some examples:
* https://github.com/wireapp/proteus (Rust implementation used in Wire app)
* https://github.com/rakoo/goax (Go implementation from Pond--probably based on an older version of the Signal protocol, but agl is known to write high-quality code)
* https://github.com/janimo/textsecure (More active Go implementation used in Ubuntu phone 3rd-party  Signal app)

**Comment by chrisballinger** [comment]
Bridging Rust and Go to iOS is much more difficult compared to C++, and it's not impossible to write relatively safe C++. The same GPL problem applies to most of those libraries, and some can even be considered derivative works if they are similar enough to Moxie's upstream code, so can't be relicensed. Depending on who you ask, this includes anything compatible with what used to be called AxolotlV3.

I've been studying this issue closely for a long time and after weighing the pros/cons of pretty much every  Axolotl library's code, language, license, authors, SignalProtocol compatibility, long term support, etc I feel that Olm is likely a solid choice for us, and other non-GPL apps in the future.

**Comment by fowlslegs** [comment]
That makes sense. They could probably all be considered derivative works since I'm sure the authors weren't thinking about doing a clean room or w/e is technically necessary for licensing reasons. Can you actually explain or link me to something that discusses the finder details of the legal situation(s)? Thanks.

**Comment by chrisballinger** [comment]
Unfortunately it's spread out over various private email threads, and a few posts on HN and Twitter.

**Comment by fowlslegs** [comment]
Okay, I was asking in a more general sense. So section 6 of the GPL is apparently what makes it App Store incompatible... I would still try to follow up w/ OWS on the mailing list and just ask really simply if they'll double-license under the MPL as well since it's GPL and OSI compatible (afaik that should do it, but I'm not a lawyer). Just be nice and don't hedge your bets on it--keep working on Olm.

**Comment by chrisballinger** [comment]
They won't use MPLv2 because it's too weak on the copyleft and threatens their current business model, especially by opening the door to MPLv2 rewrites / derivative works. There may be a GPL + "additional permission" eventually released, but I have a feeling it may be after we've already moved forward with Olm. 

Another benefit to using Olm is that OMEMO can be more widespread in other apps if it isn't hindered by a GPL dependency.

**Comment by fowlslegs** [comment]
Mmm, okay. I saw MPLv2 mentioned in another thread by you as the best
option, but I guess that was proposed and shot down already.

**Comment by aaronraimist** [comment]
@chrisballinger Looks like they have used MPLv2 https://github.com/WhisperSystems/libsignal-protocol-c/commit/defc71dfaafb8eeab05db62a4bdc88d37e2ee56b

**Comment by intelfx** [comment]
@fowlslegs @chrisballinger So now there is that [additional permission](https://whispersystems.org/blog/license-update/)...

**Comment by chrisballinger** [comment]
We may be moving to `libsignal-protocol-c` due to this announcement, but I should a better idea by the end of the week. This doesn't rule out supporting Olm as well in the future to interoperate with other non-GPL apps that may adopt Olm like Monal.

**Comment by louy2** [comment]
> I should a better idea by the end of the week.
Any updates?

**Comment by chrisballinger** [comment]
Objective-C wrapper is complete.

**Comment by DreamFlasher** [comment]
What are the steps that need to be completed so that this ticket can be closed? :) And which of these steps did you achieve?

**Comment by chrisballinger** [comment]
We have completed the libsignal-protocol-c Objective-C wrapper and the
XMPPFramework OMEMO stanza parser/generator. We are working on the
application integration right now.

**Comment by DreamFlasher** [comment]
Thank you for the update!

**Comment by therob84** [comment]
@chrisballinger great news!
After you will have realeased a ChatSecure-Verison in the near future with OMEMO, should it be completely compatible with current [Conversations](https://github.com/siacs/Conversations) of @iNPUTmice (incl. OMEMO, avaters, image-upload, etc.)? 
(I hope 100% yes and would doubt you would drop the chance to have interoparability out-of-the-box between both, but I want to be sure...)

**Comment by chrisballinger** [comment]
Our goal is compatibility

**Comment by DreamFlasher** [comment]
Hi Chris, how is it going with the application integration? Thanks for an update! :)

**Comment by Asara** [comment]
I am also curious about omemo integration.  Chatsecure's lack of it is essentially the only reason for not migrating from Signal.

**Comment by chrisballinger** [comment]
Just got back from vacation! Expect to see some good progress soon

**Comment by jsmith000** [comment]
Should we expect an update this month? :)

**Comment by chrisballinger** [comment]
Yesterday we decrypted our first OMEMO message from Conversations. Encryption is pretty close as well, and if all goes to plan we should be able to send our first encrypted message later today. Took a long time to develop all the individual pieces in a modular, reusable, well documented way, and it feels great to see them fall into place.

The big blocker for release is properly choosing and displaying the crypto state (plaintext/OTR/OMEMO), and fingerprint / device management UI. There is also a concern about security risks associated with stale devices brought up in the OMEMO security audit, so we need to work with @iNPUTmice for a shared solution on that.

**Comment by therob84** [comment]
Really really awsome your work!
Good that you stayed in contact with @iNPUTmice
<https://github.com/iNPUTmice> and Anu Pokharel here and in monal-forum.
Keep on this great work! I hope it will help spread omemo into the
world. Thanks for letting us know your progress!

**Comment by therob84** [comment]
hey @chrisballinger ... how your hard work is going these days? Still satisfied with the results?
I guess you know a lot of people are eager to read some news.....maybe also a ROUGH release date for ChatSecure with OMEMO? Best regards, Robert

**Comment by iNPUTmice** [comment]
@therob84 you might want to follow `@ChatSecure` and maybe me `@iNPUTmice` on Twitter. Receiving news in push form is probably better than polling an issue tracker every couple of days for new information. The 'is it done' question contributes nothing to the actual conversation.

**Comment by chrisballinger** [comment]
Everything works as far as the OMEMO encryption itself, but we are currently fixing things we broke before we release a beta, and polishing some new UI.

Aiming for a beta release next week?

**Comment by therob84** [comment]
@chrisballinger ... As I hoped at least YOU got my intention exactly right and you found worth to spent this 20 seconds for a fast and constructive note here - thanks, very friendly!!

P.S.:...On both your twitter streams I couldn't find anything with similar news content, Daniel, which is the reason why I repeatedly misused this issue tracker...just in lack of appropriate other media. Or would you suggest to "Shut up if you can't contribute anything to the project" (which I can't at this stage, unfortunately)? (rhetorical question!)

**Comment by chrisballinger** [comment]
@therob84 unfortunately I don't update social media very often. I think it's okay to poke us occasionally, just don't go overboard.

Estimating how long software will take is one of the unsolved problems of software development: https://www.quora.com/Why-are-software-development-task-estimations-regularly-off-by-a-factor-of-2-3

**Comment by Asara** [comment]
@chrisballinger 
Thanks for the update.  I personally don't have an iOS device but am eagerly waiting for OMEMO support.  Currently there is no good alternative for Signal because of the lack of OMEMO. 

I use LibreSignal and it is no longer being updated/expires in a couple of days, and as such the next best solution is XMPP + OMEMO.  Once this is complete, we will have the best solution, and you are appreciated greatly for your work 

Thanks for the awesome work.

**Comment by DreamFlasher** [comment]
This is all about ChatSecure for iOS right? So ChatSecure for Android is dead and will never support OMEMO?

**Comment by joostrijneveld** [comment]
On Android, I would not immediately see a benefit of competing with [Conversations](https://github.com/siacs/Conversations), with which ChatSecure-iOS will ideally be fully compatible..

**Comment by hex-m** [comment]
@DreamFlasher: @chrisballinger blogged about this a year ago. https://chatsecure.org/blog/chatsecure-conversations-zom/

**Comment by DreamFlasher** [comment]
Exactly, that was a year ago, that's why I am asking, a year is a long time, anything can change :)

**Comment by DreamFlasher** [comment]
Exactly, that was a year ago, that's why I am asking, a year is a long time, anything can change :)

**Comment by DreamFlasher** [comment]
Exactly, that was a year ago, that's why I am asking, a year is a long time, anything can change :)

**Comment by chrisballinger** [comment]
ChatSecure Android is deprecated and there are no plans to change that. The code lives on as [Zom Android](https://github.com/zom/Zom-Android) but it might not be your flavor. I was originally thinking of skinning Conversations to make a new ChatSecure Android, but we don't have the resources to keep it up to date with the latest upstream, and it wouldn't really offer much beyond a different skin.

**Comment by DreamFlasher** [comment]
Thank you @chrisballinger -- I am interested in an alternative for Conversations because it is currently too difficult to use for non-tech people. The biggest hurdly is the installation, people are not willing to spend the money in the playstore for an app that nobody of their peers except me is using, and then downloading the f-droid apk, changing the settings to allow external apk, that's all too difficult.

**Comment by lovetox** [comment]
so basically your friends dont want to spend 3 dollar/euro once to chat with you?

thats a really good reason to develop a whole client that does nothing different but not costing 3 dollar/euro. i will start to develop this for you today !

**Comment by Mikaela** [comment]
> so basically your friends dont want to spend 3 dollar/euro once to chat with you?

What if these friends don't own a payment card or don't want to feed it's detailts to Google (or anywhere else by that matter) or happen to be without any financial support and need this 3€ to food?

**Comment by DreamFlasher** [comment]
@lovetox That's the developer speaking. Well, yes exactly, nobody would spend 3 dollar if there are hundreds of free clients. Clients all their friends are using. The question I need to answer them is: "Why can't you just use Whatsapp, Facebook or Telegram as everyone else? They are all encrypted now too!" Paying 3$ means friction. Friction that will slow down the spread of Conversations.

**Comment by chrisballinger** [comment]
If I were Daniel I'd make Conversations free to download but include more in-app purchases and enable them for users who downloaded the app before a certain date. However it's his app and he can do whatever works best for him.

That said, I believe that Zom Android will get OMEMO support at some point in 2017.

**Comment by amenk** [comment]
One problem is there: countries where Google payment services are not
available.

**Comment by chrisballinger** [comment]
I'd say this issue is resolved! Stay tuned for the 4.0 release.

**Comment by vanitasvitae** [comment]
That's really great news!

**Comment by Asara** [comment]
Thank you so much!

**Comment by AtosNicoS** [comment]
Chatsecure is not listed as fully supported in this chart:
http://omemo.top

Is this information out of date or what is missing in chatsecure? Does omemo work for a group chat?

**Comment by hermann-san** [comment]
getting the OMEMO list changed was requested  in Oct.  2019 
https://github.com/bascht/omemo-top/issues/189

**Comment by tuxayo** [comment]
I'm not sure what should be done. Opening a new issue to request updating the page?
What is the support for groups chats? So I can give the info.

**Comment by licaon-kter** [comment]
@tuxayo Group OMEMO is supported >2 years, what do you mean?
